### ISSUE TITLE
Script Engine Crash

### ISSUE THREAD

**Issue by Holzhaus** [Holzhaus/mixxx-gh-issue-migration#320]
Some intermitent crashes with Mixxx 1.7.1 (full system information is in the attached file).
I cannot find a reliable way to reproduce, the problem seems to be completely random and happens more or less once per hour.

Here is the gdb log/backtrace:

Debug: [Main]: Load to player2: "/home/extra/Musik/cadenza/[cadenza 009] luciano & salif keita - salif keita remix/01 - Luciano & Salif Keita - Yamore (Luciano Remix).mp3" 
Debug: [Reader 2]: file length  65889792 i 
Debug: [AnalyserQueue 1]: AnalyserWaveform: f  44100  samplesPerDownsample:  110  downsamples  598998  from  65889792 
Debug: [Main]: Received waveform from track 

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb1ae1b70 (LWP 5545)]
0xb68a0c1c in ?? () from /lib/i686/cmov/libc.so.6
(gdb) bt
#0  0xb68a0c1c in ?? () from /lib/i686/cmov/libc.so.6
#1  0xb68a317e in malloc () from /lib/i686/cmov/libc.so.6
#2  0xb6c0fe1d in qMalloc(unsigned int) () from /usr/lib/libQtCore.so.4
#3  0xb6c5bcd9 in QString::realloc(int) () from /usr/lib/libQtCore.so.4
#4  0xb6e7ee7e in ?? () from /usr/lib/libQtScript.so.4
#5  0xb6e81856 in ?? () from /usr/lib/libQtScript.so.4
#6  0xb6e772ca in ?? () from /usr/lib/libQtScript.so.4
#7  0xb6e81856 in ?? () from /usr/lib/libQtScript.so.4
#8  0xb6e968cb in ?? () from /usr/lib/libQtScript.so.4
#9  0xb6ed1e67 in QScriptValue::call(QScriptValue const&, QList<QScriptValue> const&) () from /usr/lib/libQtScript.so.4
#10 0x081cbe2d in ?? ()
#11 0x081cbfa9 in ?? ()
#12 0x080e3112 in ?? ()
#13 0x081d3240 in ?? ()
#14 0xb6c155e2 in ?? () from /usr/lib/libQtCore.so.4
#15 0xb6ab4585 in start_thread () from /lib/i686/cmov/libpthread.so.0
#16 0xb69022be in clone () from /lib/i686/cmov/libc.so.6

**Comment by Holzhaus** [comment]
(by gimmeapill)

**Comment by Holzhaus** [comment]
(by wesley-werner)
Greetings. 
I'm not sure if our crashes are related, they seem to fail at the same point, but I get different error, are you running a compiled version perhaps?. I'm using Mixxx 1.7.1 from the download page, running on Ubuntu Karmic, kernel 2.6.31-15. 

Note I'm aware of the PortAudio issue for Ubuntu 9.04, but uncertain of it's status in 9.10. The crash happens before I play any audio, for ALSA, OSS or even when None device is selected.

I can recreate this crash by loading a specific file into "Player 1" (from a fresh Mixx startup, nothing playing). The file has 0 BPM and loads the waveform display for a split second, and then it bombs out. Most of the other files load fine.

Let me know if there's any way I can provide more info, willing to compile if needed.

Attached is the backtrace.

**Comment by Holzhaus** [comment]
(by gamegod)
Wesley: Can you try to reproduce the crash while running Mixxx under gdb?

A backtrace including information for all of our threads would be very
helpful, and we have instructions for getting that here:
http://mixxx.org/wiki/doku.php/creating_backtraces

I have the feeling Mixxx is bombing in the BPM detection algorithm in
both of your cases. Can you check to see if there's specific files
that are causing reproducible crashes?

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Wesley, 
I cannot tell for sure if it is the same bug, but I could reproduce both with the official debian package, the non official ubuntu package and my compiled binaries. It doesn't seem to happen particularly when loading a song, but I may be wrong. I'm trying to reproduce the problem at the moment under gdb, but didn't get any good backtrace yet...

**Comment by Holzhaus** [comment]
(by wesley-werner)
Hi, I will run and post the gdb results tonight after work. Thanks for the
link Albert.

**Comment by Holzhaus** [comment]
(by wesley-werner)
Hi again. Okay here is the output for gdb mixxx, I did notice that dbg didn't find any symbols. I'll happily do this with symbols, I'm not sure what that involves though; Just compiling Mixxx myself with debug options, and running that through gdb?

**Comment by Holzhaus** [comment]
(by gamegod)
Hi Wesley,

Thanks for the more detailed backtrace. If you compile Mixxx from
scratch, debugging symbols will be enabled by default.

At the bottom of your output, GDB reported this:
---Type <return> to continue, or q <return> to quit---

Unfortunately, it only showed a backtrace for our first 3 threads -
GDB wanted you to hit enter a few more times to scroll the terminal
and show you the rest of the backtrace. This is an easy mistake to
make. :) Do you think you can try to reproduce the crash one more time
and grab the rest of gdb's output?

You've been a great help so far, it's very much appreciated.

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by wesley-werner)
Ah silly me, first time using gdb ;)

Okay I got all the deps sorted, compiled, and did a bt with the symbols loaded.

I'll create some sound files and see if one will recreate this behavior, but time is against me until the weekend. I'm playing a live set with Mixxx tomorrow night, actually (I will avoid the faulty song for now, no problem).

Thanks for the work BTW, much appreciated!

**Comment by Holzhaus** [comment]
(by gimmeapill)
Yay! Got another one after almost 4 hours!
I copied the full gdb session + backtrace this time, I hope this will be enough...

**Comment by Holzhaus** [comment]
(by gamegod)
Looks like you've identified a race condition in our script engine.
We'll definitely have to take a look at this and see if we can track
it down. Big help guys, thanks!

Albert

**Comment by Holzhaus** [comment]
(by gamegod)
Hey gimmeapill, in Mixxx's preferences, you can disable the BPM detection. Can you try doing that and seeing if you get any crashes?

Because Mixxx is crashing inside a new/malloc, it's possible there's some sort of subtle memory corruption going on. If that's the case, then oddball bits of code like the BPM detector could be stepping on the script engine's memory. (Make sure you test under gdb again, just in case.)

Another test you can do is switching from pitch-independent time stretch to vinyl emulation mode or vice versa. Please test with the BPM detection turned off before you try this though...

Thanks,
Albert

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Albert,
I never use BPM detection or pitch-independant timestrech. Actually my library has not even been scanned by the BPM detector. But I don't think the problem comes from here as I've just run a full night gdb session on another computer with the same library playing in loop, but without the Hercules RMX and it didn't crash.
So I'll run another one today with the RMX connected and the pitch-independant timestrech see what happens..

**Comment by Holzhaus** [comment]
(by gamegod)
For reference, Msieur Piero on the forums has a backtrace that's very similar to gimmeapill's backtrace (crash in memory allocation from QScriptValue::call()), and he's also using the Hercules RMX:

http://mixxx.org/forums/viewtopic.php?f=3&t=967&st=0&sk=t&sd=a

**Comment by Holzhaus** [comment]
(by gamegod)
Ok, someone humour me. I spotted a bit of code in the Hercules RMX script that might be "stressing" the script engine in some way. 

Crack open your Hercules-DJ-Console-RMX-scripts.js in your Mixxx/midi folder and edit lines 33 and 34, and change:

 33     engine.connectControl("[Channel1]","playposition","HerculesRMX.wheelDecay");  
 34     engine.connectControl("[Channel2]","playposition","HerculesRMX.wheelDecay");   

to:

 33     //engine.connectControl("[Channel1]","playposition","HerculesRMX.wheelDecay");  
 34     //engine.connectControl("[Channel2]","playposition","HerculesRMX.wheelDecay");   

In other words, just commenting them out. The wheelDecay() function in the script uses the Date class in a way that the Stanton SCS.3d script doesn't (Hercules-DJ-Console-RMX-scripts.js line 41). By commenting out the above connectControl() calls, we should prevent the wheelDecay() function from ever being called. (I'll save my vague speculation as to how this could be causing a problem for later.)

Let's see if commenting out those lines makes a difference gimmeapill, using your original configuration. I'd appreciate it if anyone else using a Hercules RMX could test it as well. 

Thanks for your help,
Albert

**Comment by Holzhaus** [comment]
(by pierre-blanco)
I tried to uncomment the lignes 33 and 34 in your script...
It causes me two surprises, one good, one bad:
- First one : the jogwel was still working (not the scratch mode and the librairies search mode, but the beatmatching still work),
- Second one : the mixxxing time increase a bit, but still kills himself after 1H30 (I gain 30mn...)

here's the backtrace, if it can helps :

Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb13d7b90 (LWP 5088)]
0xb6961a44 in ?? () from /lib/tls/i686/cmov/libc.so.6
(gdb) thread apply all bt

Thread 16 (Thread 0xadc22b90 (LWP 5093)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb69c9ae7 in poll () from /lib/tls/i686/cmov/libc.so.6
#2  0xb6b9c2fb in ?? () from /usr/lib/libportaudio.so.2
#3  0xb6ba1124 in ?? () from /usr/lib/libportaudio.so.2
#4  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#5  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 15 (Thread 0xae423b90 (LWP 5091)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0xce5ea30, mutex=0xce5ea38, time=4294967295) at thread/qwaitcondition_unix.cpp:87
#3  0x081c1755 in ?? ()
#4  0xb7c0396e in QThreadPrivate::start (arg=0xce5e9b8) at thread/qthread_unix.cpp:189
#5  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#6  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 14 (Thread 0xafffeb90 (LWP 5090)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0xcf7ab40, mutex=0xcf7ab48, time=4294967295) at thread/qwaitcondition_unix.cpp:87
#3  0x081c1755 in ?? ()
#4  0xb7c0396e in QThreadPrivate::start (arg=0xcf7aac8) at thread/qthread_unix.cpp:189
#5  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#6  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 13 (Thread 0xb1bd8b90 (LWP 5089)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb69c9ae7 in poll () from /lib/tls/i686/cmov/libc.so.6
#2  0xb66ed74b in g_poll () from /usr/lib/libglib-2.0.so.0
#3  0xb66dff82 in ?? () from /usr/lib/libglib-2.0.so.0
#4  0xb66e0268 in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#5  0xb7d23438 in QEventDispatcherGlib::processEvents (this=0xd068c00, flags={i = -1312980376}) at kernel/qeventdispatcher_glib.cpp:323
#6  0xb7cf606a in QEventLoop::processEvents (this=0xb1bd82e0, flags={i = -1312980312}) at kernel/qeventloop.cpp:149
#7  0xb7cf64aa in QEventLoop::exec (this=0xb1bd82e0, flags={i = -1312980248}) at kernel/qeventloop.cpp:200
#8  0xb7c00639 in QThread::exec (this=0xb0a19008) at thread/qthread.cpp:481
#9  0x081d1a31 in ?? ()
#10 0xb7c0396e in QThreadPrivate::start (arg=0xb0a19008) at thread/qthread_unix.cpp:189
#11 0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
---Type <return> to continue, or q <return> to quit---
#12 0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 12 (Thread 0xb13d7b90 (LWP 5088)):
#0  0xb6961a44 in ?? () from /lib/tls/i686/cmov/libc.so.6
#1  0xb69639c5 in malloc () from /lib/tls/i686/cmov/libc.so.6
#2  0xb7bfe0bd in qMalloc (size=31) at global/qmalloc.cpp:55
#3  0xb7c06a58 in QByteArray::realloc (this=0xb0bd8138, alloc=11) at tools/qbytearray.cpp:1379
#4  0xb7cff9cf in normalizeTypeInternal (t=0xb0bd8180 "MidiMessage", e=0xb0bd818b "", fixScope=false, adjustConst=true)
    at ../../include/QtCore/../../src/corelib/tools/qbytearray.h:424
#5  0xb7d000bd in qNormalizeType (d=0xb0bd8180 "MidiMessage", templdepth=<value optimized out>, result=@0xb0bd82c8) at kernel/qmetaobject.cpp:917
#6  0xb7d01684 in QMetaObject::normalizedType (type=0x81ea463 "MidiMessage") at kernel/qmetaobject.cpp:947
#7  0xb7d01c9f in QMetaType::registerType (typeName=0x81ea463 "MidiMessage", destructor=0x80fa400, constructor=0x80fa420) at kernel/qmetatype.cpp:426
#8  0x080f8b30 in ?? ()
#9  0x080e4647 in ?? ()
#10 0x081da2b0 in ?? ()
#11 0xb7c0396e in QThreadPrivate::start (arg=0xd46ea58) at thread/qthread_unix.cpp:189
#12 0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#13 0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 10 (Thread 0xb2465b90 (LWP 5086)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0x9fe7ff8, mutex=0x9fe7ff4, time=4294967295) at thread/qwaitcondition_unix.cpp:87
#3  0x080dec2a in ?? ()
#4  0x080ded73 in ?? ()
#5  0xb7c0396e in QThreadPrivate::start (arg=0x9fe7fe0) at thread/qthread_unix.cpp:189
#6  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#7  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 9 (Thread 0xaec24b90 (LWP 5085)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb63ea78f in ?? () from /usr/lib/libjack.so.0
#3  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#4  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 4 (Thread 0xb3599b90 (LWP 5076)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0x9fc1410, mutex=0x9fc140c, time=4294967295) at thread/qwaitcondition_unix.cpp:87
---Type <return> to continue, or q <return> to quit---
#3  0x080dceed in ?? ()
#4  0xb7c0396e in QThreadPrivate::start (arg=0x9fc13e0) at thread/qthread_unix.cpp:189
#5  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#6  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 3 (Thread 0xb443eb90 (LWP 5075)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0x9f5f918, mutex=0x9f5f91c, time=4294967295) at thread/qwaitcondition_unix.cpp:87
#3  0x08135a7d in ?? ()
#4  0xb7c0396e in QThreadPrivate::start (arg=0x9f5f8d8) at thread/qthread_unix.cpp:189
#5  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#6  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 2 (Thread 0xb515eb90 (LWP 5074)):
#0  0xb7f93430 in __kernel_vsyscall ()
#1  0xb6b820e5 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/tls/i686/cmov/libpthread.so.0
#2  0xb7c049b2 in QWaitCondition::wait (this=0x9f6eeb8, mutex=0x9f6eebc, time=4294967295) at thread/qwaitcondition_unix.cpp:87
#3  0x08135a7d in ?? ()
#4  0xb7c0396e in QThreadPrivate::start (arg=0x9f6ee78) at thread/qthread_unix.cpp:189
#5  0xb6b7e4ff in start_thread () from /lib/tls/i686/cmov/libpthread.so.0
#6  0xb69d449e in clone () from /lib/tls/i686/cmov/libc.so.6

Thread 1 (Thread 0xb618f710 (LWP 5073)):
#0  0xb2865326 in drm_intel_bo_unreference () from /usr/lib/libdrm_intel.so.1
#1  0xb28c6bcd in brw_draw_prims () from /usr/lib/dri/i965_dri.so
#2  0xb2980dd0 in ?? () from /usr/lib/dri/i965_dri.so
#3  0xb7174c74 in QOpenGLPaintEngine::drawLines (this=0xcadfb50, lines=0x0, lineCount=1) at qpaintengine_opengl.cpp:3974
#4  0xb743457a in QPainter::drawLines (this=0xbf82c70c, lines=0xbf82c5e0, lineCount=1) at painting/qpainter.cpp:4568
#5  0x0816d5e1 in ?? ()
#6  0x08166e3b in ?? ()
#7  0x08164263 in ?? ()
#8  0xb735691e in QWidget::event (this=0xa33db48, event=0xbf82cbc4) at kernel/qwidget.cpp:7654
#9  0xb714d99c in QGLWidget::event (this=0xa33db48, e=0xbf82cbc4) at qgl.cpp:3235
#10 0xb72ffe9c in QApplicationPrivate::notify_helper (this=0x9d89340, receiver=0xa33db48, e=0xbf82cbc4) at kernel/qapplication.cpp:4084
#11 0xb7308282 in QApplication::notify (this=0x9d89330, receiver=0xa33db48, e=0xbf82cbc4) at kernel/qapplication.cpp:4049
#12 0xb7cf7a3b in QCoreApplication::notifyInternal (this=0x9d89330, receiver=0xa33db48, event=0xbf82cbc4) at kernel/qcoreapplication.cpp:602
#13 0xb730afee in QCoreApplication::sendSpontaneousEvent (receiver=0xa33db48, event=0x0)
    at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:216
#14 0xb735e719 in QWidgetPrivate::drawWidget (this=0xa33df60, pdev=0xa33db50, rgn=@0xbf82ccf8, offset=@0xbf82ccf0, flags=3, sharedPainter=0x0, 
---Type <return> to continue, or q <return> to quit---
    backingStore=0x0) at kernel/qwidget.cpp:5046
#15 0xb7519668 in QWidgetPrivate::repaint_sys (this=0xa33df60, rgn=@0xa33e040) at painting/qbackingstore.cpp:1534
#16 0xb734eb5e in QWidgetPrivate::syncBackingStore (this=0xa33df60) at kernel/qwidget.cpp:1600
#17 0xb7356d85 in QWidget::event (this=0xa33db48, event=0xb59b5f8) at kernel/qwidget.cpp:7794
#18 0xb714d99c in QGLWidget::event (this=0xa33db48, e=0xb59b5f8) at qgl.cpp:3235
#19 0xb72ffe9c in QApplicationPrivate::notify_helper (this=0x9d89340, receiver=0xa33db48, e=0xb59b5f8) at kernel/qapplication.cpp:4084
#20 0xb7308282 in QApplication::notify (this=0x9d89330, receiver=0xa33db48, e=0xb59b5f8) at kernel/qapplication.cpp:4049
#21 0xb7cf7a3b in QCoreApplication::notifyInternal (this=0x9d89330, receiver=0xa33db48, event=0xb59b5f8) at kernel/qcoreapplication.cpp:602
#22 0xb7cf8695 in QCoreApplicationPrivate::sendPostedEvents (receiver=0x0, event_type=0, data=0x9d88460)
    at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:213
#23 0xb7cf888d in QCoreApplication::sendPostedEvents (receiver=0x0, event_type=0) at kernel/qcoreapplication.cpp:1132
#24 0xb7d237ef in postEventSourceDispatch (s=0x9d91190) at ../../include/QtCore/../../src/corelib/kernel/qcoreapplication.h:218
#25 0xb66dcb88 in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#26 0xb66e00eb in ?? () from /usr/lib/libglib-2.0.so.0
#27 0xb66e0268 in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#28 0xb7d23438 in QEventDispatcherGlib::processEvents (this=0x9d8eeb0, flags={i = -1081945256}) at kernel/qeventdispatcher_glib.cpp:323
#29 0xb73a1365 in QGuiEventDispatcherGlib::processEvents (this=0x9d8eeb0, flags={i = -1081945208}) at kernel/qguieventdispatcher_glib.cpp:202
#30 0xb7cf606a in QEventLoop::processEvents (this=0xbf82d400, flags={i = -1081945144}) at kernel/qeventloop.cpp:149
#31 0xb7cf64aa in QEventLoop::exec (this=0xbf82d400, flags={i = -1081945080}) at kernel/qeventloop.cpp:200
#32 0xb7cf8959 in QCoreApplication::exec () at kernel/qcoreapplication.cpp:880
#33 0xb72ffd17 in QApplication::exec () at kernel/qapplication.cpp:3553
#34 0x080e2f40 in ?? ()
#35 0xb6906775 in __libc_start_main () from /lib/tls/i686/cmov/libc.so.6
#36 0x080649e1 in ?? ()
#0  0xb6961a44 in ?? () from /lib/tls/i686/cmov/libc.so.6

**Comment by Holzhaus** [comment]
(by gamegod)
Hi Msieur Piero (and gimmeapill),

Thanks for testing that for me. Today, three of us put our heads together and fixed some small logic errors in our analyser code. I think there's a decent chance this may have fixed your problem.

As soon as you can, can you please check out our latest 1.7 branch code (it includes the bug fixes) and test once again:
  bzr co lp:mixxx/1.7 mixxx_170
  cd mixxx_170
  scons
  ./mixxx

You can uncomment those two lines in the Hercules again, since I don't think they caused the problem. 

Thank you very much!
Albert

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Albert,
I just gave it a try and the latest fixes seem to have caused a huge performance hit. 
While I could run smoothly at 12ms with 1.7.1, the latest bzr is barely usable: sound is choppy even at 64ms, and the waveform dsiplay stutters as well. I verified by re-compiling 1.7.1 with the same settings (sudo scons tuned=1 install).
As for the crash it didn't happen again so far. Msieur Piero, can you confirm?

**Comment by Holzhaus** [comment]
(by adamdavison)
Hi,

Albert is this your power-of-two buffer size patch in action? Can't
believe that analyser change would have caused a performance hit...

Adam

2009/12/23 gimmeapill <email address hidden>:
> Hi Albert,
> I just gave it a try and the latest fixes seem to have caused a huge performance hit.
> While I could run smoothly at 12ms with 1.7.1, the latest bzr is barely usable: sound is choppy even at 64ms, and the waveform dsiplay stutters as well. I verified by re-compiling 1.7.1 with the same settings (sudo scons tuned=1 install).
> As for the crash it didn't happen again so far. Msieur Piero, can you confirm?
>
> --
> Script Engine Crash
> https://bugs.launchpad.net/bugs/493793
> You received this bug notification because you are a member of Mixxx
> Development Team, which is subscribed to Mixxx.
>

**Comment by Holzhaus** [comment]
(by pierre-blanco)
I've tried several times to type :
bzr co lp/mixxx/1.7 mixxx_170

but the recording always fails at 500 or 600/2589...
Do I forget something important?

I'm quite a noob with the bzr stuff, I know...

**Comment by Holzhaus** [comment]
(by gamegod)
gimmeapill: I've just committed even more new code to rework the latency calculation. Please try again and see if you have any performance problems. (Make sure you have optimizations turned on!) :)

Msieur Piero: Make sure there is a ":" and not a "/" after "lp". The command should be:

   bzr co lp:mixxx/1.7 mixxx_170

It's possible that our Bzr hosting on Launchpad was down yesterday. Please try again!

Thanks again guys,
Albert

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Albert,

With the latest fixes, performance is back to an OK level (23ms without glitches) although there's still a sensible performance gap compared to the  Debian package (http://packages.debian.org/unstable/sound/mixxx).
I still cannot explain why there should be a performance difference, but the Debian binaries run definitely smoother on my system than the ones I produce, (even with optimizations turned on and a stripped binary). It is especially visible with the waveform display. This is anyway unrelated to the script engine bug.

As for the segfault problem, it doesn't seem to happen so far, but I would need a bit more testing, as this is quite random.

Msieur Piero:
here's the short script I use, if that can be of any help (this requires that you are ini the sudoers/wheel group):

#!/bin/sh

# delete the build folder
sudo rm -r 1.7

# get the latest bzr
bzr checkout lp:mixxx/1.7 1.7

# Rename the old binary
sudo mv /usr/local/bin/mixxx /usr/local/bin/mixxx-`date +%Y%m%d%H%M`

# compile & install
sudo scons tuned=1 install

**Comment by Holzhaus** [comment]
(by pierre-blanco)
Sorry for my late answering...  My in-laws broadcast was such a mess so I couldn't download the whole bzr stuff...
But...
A fiew asks are still there. 
- Do I have to uninstall mixxx 1.7.1 BEFORE using scons or do I have to scons on the existing install?
- I have nothing under usr/local/bin/mixxx
- When I use scons into ./mixxx_170 It shows scons: 
*** No SConstruct file found.
File "/usr/lib/scons/SCons/Script/Main.py", line 830, in _main

Into /home/pierre/mixxx_170/mixxx/ scons tuned=1 install shows :
pierre@pierre-laptop:~/mixxx_170/mixxx$ scons tuned=1 install
scons: Reading SConscript files ...
Platform: Linux i686
WE ARE IN: /home/pierre/mixxx_170/mixxx/src
QT path: /usr/share/qt4
Loading qt4 tool...
Exception: Qt4 command 'moc' not found. Tried: /usr/share/qt4/bin/moc-qt4 and /usr/share/qt4/bin/moc:
  File "/home/pierre/mixxx_170/mixxx/SConstruct", line 42:
    SConscript(File('src/SConscript'), build_dir=Dir(build_dir), duplicate=0)
  File "/usr/lib/scons/SCons/Script/SConscript.py", line 612:
    return apply(method, args, kw)
  File "/usr/lib/scons/SCons/Script/SConscript.py", line 549:
    return apply(_SConscript, [self.fs,] + files, subst_kw)
  File "/usr/lib/scons/SCons/Script/SConscript.py", line 259:
    exec _file_ in call_stack[-1].globals
  File "/home/pierre/mixxx_170/mixxx/src/SConscript", line 172:
    env = Environment(tools=['default','qt4'], toolpath=['#build/'], QTDIR=flags_qtdir, ENV = os.environ)
  File "/usr/lib/scons/SCons/Environment.py", line 992:
    apply_tools(self, tools, toolpath)
  File "/usr/lib/scons/SCons/Environment.py", line 106:
    env.Tool(tool)
  File "/usr/lib/scons/SCons/Environment.py", line 1676:
    tool(self)
  File "/usr/lib/scons/SCons/Tool/__init__.py", line 181:
    apply(self.generate, ( env, ) + args, kw)
  File "/home/pierre/mixxx_170/mixxx/build/qt4.py", line 251:
    QT4_MOC = locateQt4Command(env,'moc', env['QTDIR']),
  File "/home/pierre/mixxx_170/mixxx/build/qt4.py", line 234:
    raise Exception("Qt4 command '" + command + "' not found. Tried: " + fullpath1 + " and "+ fullpath2)

Does everything seems alright?

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Msieur Piero,

Do I have to uninstall mixxx 1.7.1 BEFORE using scons or do I have to scons on the existing install?
-> if you mean the ubuntu package, then yes. The one you compile will go by default under /usr/local/bin so this is not 100% mandatory but otherwise you might run the wrong binary.

- I have nothing under usr/local/bin/mixxx
-> you need root privileges to run the install command, so you must "su" first or run "sudo scons tuned=1 install"

- When I use scons into ./mixxx_170 It shows scons:....
-> I'm not too sure, but I would say you miss some build dependencies (like libqt4-dev)
Be sure to check the build instructions in the wiki: http://mixxx.org/wiki/doku.php/compiling_on_linux

if everything goes fine, you will end up with the Mixxx binary under /usr/local/bin, and the rest under /usr/local/share/mixxx.

**Comment by Holzhaus** [comment]
(by pierre-blanco)
I got a slight problem with the libqt4-dev dependancies...
I'm gonna check it first, and after, i'll check the scons...

**Comment by Holzhaus** [comment]
(by rryan)
- When I use scons into ./mixxx_170 It shows scons:
*** No SConstruct file found.
File "/usr/lib/scons/SCons/Script/Main.py", line 830, in _main
--------

You need to use the scons command in ./mixxx_170/mixxx/, not ./mixxx_170  :)


-RJ Ryan

**Comment by Holzhaus** [comment]
(by pierre-blanco)
Ok...
Everything went well... I succed in building mixxx from source (libqt4-dev was such a nightmare to install, but I went through it).
To apologize, I must say it was the first "big software" I built (I used scons for little stuff like taputape, or some other tiny soft).

I begin to test it yesterday (i've not so much time), but, with the little i mixxxed, i would say that it looks like the improvements you made seems to work.

I'm gonna test it deeper tonight.

Thanks everybody.

Keep in touch!

**Comment by Holzhaus** [comment]
(by rryan)
Not sure why this was marked milestone 1.8.0. Ideally this was fixed in 1.7.2. Msiuer Piero, gimmeapill: can you confirm that the problem is resolved? You can download 1.7.2 from our website now.

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Rj, yes, this is correct. 
I've been running 1.7.2 since it came out, but I didn't run it for long enough to be sure the crash condition is gone 
(Mixxx needs to run at least for 1-2 hours while being actively used). 
This is why I didn't update the bug report yet but it does look better.
I'll run some long sessions this week end.

**Comment by Holzhaus** [comment]
(by gimmeapill)
Hi Rj, Albert,

The crash condition is still not fixed with 1.7.2. 
Performance improved but Mixxx still segfaults after about 1 hour of playing actively with the RMX.
I'll try to generate another gdb trace ASAP...

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
$10 says that the threading issue I fixed in the features_scriptTimers branch is to blame for this as well. Shall I back-port?

**Comment by Holzhaus** [comment]
(by gimmeapill)
That would be sweet :-)
But maybe to save you some time I should rather try directly your branch? is it usable/stable enough?

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
It is. Please try it and let me know.

**Comment by Holzhaus** [comment]
(by gimmeapill)
OK, after a few hours playing with your branch, the crash iself seems to be gone but new problems appeared (I guess they're related to 1.8 as trunk behaves more or less the same): lights not working anymore after a while, midi controls acting weirdly, small audio glitches when loading a track. Since this renders things hard to compare, would it be possible for to backport your fixes to the 1.7 branch?

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
It's possible to back-port, but might take more time than just waiting for 1.8 to include this (which may not happen until 1.8.1, depends on what the head devs say.)
The rest of those issues are related to using 1.8 over 1.7 except the lights-not-working one. Does that sound like this bug? https://bugs.launchpad.net/mixxx/+bug/410750 Like does the controller stop working altogether or just the lights?

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
Marking in-progress until a decision is made on merging features_scriptTimers or porting the fixes to trunk.

**Comment by Holzhaus** [comment]
(by gimmeapill)
By the way, the latest 1.7.x is way more stable now. 
It still crashes, but only after about 2.5 hours of playing (it used to be roughly every hour), so the latest fixes definitely helpded...

**Comment by Holzhaus** [comment]
(by pegasus-renegadetech)
features_scriptTimers has been merged to trunk. Marking fixed.
